### DISCUSSION TITLE
How to register 'react-bootstrap-typeahead' component using React 'useForm' hook?

### DISCUSSION THREAD

**Discussion by haizendaizen** [react-hook-form/react-hook-form#2624] [Q&A] [answered]
I'm building a simple html form using "react-hook-form" library: https://react-hook-form.com/

I've incorporated "react-bootstrap-typeahead" into the html form but haven't been able to register this component with 'useForm' hook. Hence, "react-bootstrap-typeahead" input data is ignored during onSubmit.

"react-bootstrap-typeahead" [doesn't provide a "name" prop](https://github.com/ericgio/react-bootstrap-typeahead/blob/master/docs/API.md#typeahead) which makes it difficult to register the component.

I've read the 'useForm' documentation on the different options for registering this type of components but still don't understand how to achieve this: https://react-hook-form.com/get-started#Registerfields

Does anybody have faced such challenge before?

It would be great to see a working example to get a better idea on how to implement "react-bootstrap-typeahead" + "react-hook-form" in my application. Thanks!

Here's my sample code:
```
import useForm from 'react-hook-form';
import { Typeahead } from 'react-bootstrap-typeahead';
import 'react-bootstrap-typeahead/css/Typeahead.css';

const myForm = (props) => {

    const { register, handleSubmit, errors } = useForm();

    const onSubmit = data => {
          // api post request with form data
        })
    };

    const mydata = [ "one", "two", "three" ];

   return (
   <>
   <form onSubmit={handleSubmit(onSubmit)} >
   <div className="form-group">

   {/* Here I'm registering text input using ref: */}

   <input type="text" className="form-control" name="name" ref={register({ required: true })} />
                                        
   </div>
                                                             
   <div className="form-group mb-0">

   {/* How can I register the below component with useForm? */}
   <Typeahead
      id="multiple-typeahead"
      clearButton
      multiple
      options={mydata}
   />
   </div>
   <button type="submit">Save</button>
   </form>
   </>
  );
}
```

**Comment by bluebill1049** [comment, marked as answer]
I think you are after: https://react-hook-form.com/api#Controller

**Reply by haizendaizen** [reply]
Hey @bluebill1049 - I gave another try at the documentation and the above lecture provided great info for this implementation. Specially the videos section! I'm enjoying working with react-hook-form, it's really flexible and makes things easier for us. Thanks for sharing this library with everyone.

Here's the code snippet of how I was able to register the component, just in case someone needs it:

```
import useForm from 'react-hook-form';
import { useForm, Controller } from "react-hook-form";
import 'react-bootstrap-typeahead/css/Typeahead.css';

const myForm = (props) => {

    const { register, handleSubmit, errors, control } = useForm();

    const onSubmit = data => {
          // api post request with form data
        })
    };

    const mydata = [ "one", "two", "three" ];

   return (
   <>
   <form onSubmit={handleSubmit(onSubmit)} >
   <div className="form-group">
   <input type="text" className="form-control" name="name" ref={register({ required: true })} />
   </div>
                                                             
   <div className="form-group mb-0">
<Controller
   as={Typeahead}
             control={control}
             name="typeahead_component"
             rules={{ required: true }}
             id="multiple-typeahead"
             clearButton
             multiple
             options={mydata}
             defaultValue=""
/>
   </div>
   <button type="submit">Save</button>
   </form>
   </>
  );
}
```

**Reply by bluebill1049** [reply]
Thanks for sharing the solution! I am glad it's working for you and everyone else.

**Comment by gdavis321** [comment]
Thanks haizendaizen, very helpful!  But I could not get  rules={{ required: true }} to work.  What I ended up doing was using a watch to populate a hidden field and added  rules={{ required: true }} to the hidden input .... any suggestions?

**Comment by Moshyfawn** [comment]
@gdavis321, using RHF V7, here's how you can integrate RBT with RHF:
```jsx
<Controller
  control={control}
  name="typeahead"
  rules={{
    required: "Please, select at least one Typeahead input value"
  }}
  render={({ field, fieldState }) => (
    <div className="mb-3">
      <label htmlFor="typeahead" className="form-label">
        Typeahead
      </label>
      <Typeahead
        {...field}
        id="typeahead"
        multiple
        clearButton
        className={fieldState.invalid ? "is-invalid" : ""}
        aria-describedby="typeaheadError"
        options={options}
      />
      <p id="typeaheadError" className="invalid-feedback">
        {fieldState.error?.message}
      </p>
    </div>
  )}
/>
```

See [this CSB](https://codesandbox.io/s/pedantic-faraday-cjdntp?file=/src/App.js:1192-1206) for a full example

**Reply by gdavis321** [reply]
thanks!

**Reply by gdavis321** [reply]
Moshyfawn,  I understand a little about the design principle of controlled and uncontrolled components, but it would be very helpful if there was a way to have something like react-hook-form's setValue for uncontrolled type-aheads.  I had a request to have the same type-ahead on 2 different tabs and each would drive the other but because I couldn't use a watch with setValue,  I ended up just displaying the value selected on one tab and had the regular type-ahead on the other tab.  I didn't try using a controlled type-ahead bc all the other fields were uncontrolled (and the form has over 70 fields so I ran into performance issues when using controlled fields).    I am a React newbie, so I apologize if I am missing basic stuff.
